diracx-db/src/diracx/db/sql/utils/functions.py

```diff
@@ -140,3 +146,53 @@ def substract_date(**kwargs: float) -> datetime:
 
 def hash(code: str):
     return hashlib.sha256(code.encode()).hexdigest()
+
+
+class DBStateAssertion:
+    """Context manager to ensure the integrity of a database transaction,
+    specifically by rolling back any changes if certain expected exceptions
+    occur—and raising a standardized error to indicate the database is in a bad state.
+
+    Example:
+    with DBStateAssertation(self.conn, [PilotNotFoundError, SecretNotFoundError]):
+        # Some code that should not raise any error filled previously
+        # Else, raises an error
+
+    Note: Will rollback changes.
+
+    """
+
+    def __init__(
+        self, conn: AsyncConnection, exceptions: list[type[Exception]]
+    ) -> None:
+        self.exceptions = exceptions
+        self.conn = conn
+
+    async def __aenter__(self):
+        pass
+
+    async def __aexit__(self, exc_type, exc_value, exc_tb):
+        if exc_type is None:
+            # No exception occurred
+            return False
+
+        # If we get here, an error occurred, so we rollback changes
+        message = (
+            f"Rolling back DB transaction due to: {exc_type.__name__}: {exc_value}"
+        )
+        await self.conn.rollback()
```

Comment:
I guess you don't need to rollback manually here, as it's supposed to be done in https://github.com/DIRACGrid/diracx/blob/a1e542c97ab1b297259e4e471249fee4c04a77d9/diracx-db/src/diracx/db/sql/utils/base.py#L208-L217
 
```suggestion
```